treewide: use `mkPackageOption`

h7x4 opened the pull request:
## Description of changes

This commit replaces a lot of usages of `mkOption` with the package type, to be `mkPackageOptionMD`, in order to reduce the amount of code.

I haven't changed all of these, but this is a good amount of them. Maybe we could just start with this, and do another PR later? I might add on stuff as time goes if it doesn't seem to get merged anytime soon.

Also, I haven't used any tool for this, I've just manually gone through them all. There might be some human errors due to that.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
  **[Module Updates](https://github.com/NixOS/nixpkgs/blob/master/nixos/README.md#module-updates)**
  - [x] changes are backward compatible
  - [x] removed options are declared with `mkRemovedOptionModule`
  - [x] changes that are not backward compatible are documented in release notes
  - [ ] ~module tests succeed on ARCHITECTURE~ This is probably too large for me to build, not entirely sure how I would go about doing this.
  - [x] options types are appropriate
  - [x] options description is set
  - [ ] ~options example is provided~ I've transferred the examples where the option already had one
  - [x] documentation affected by the changes is updated

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

wegank commented:
Why not just `mkPackageOption`?

https://github.com/NixOS/nixpkgs/blob/96ae9c987800d68911ae7fe50e088f2f2ad46827/lib/options.nix#L226

h7x4 commented:
@wegank Mostly for ease of backporting, at least for however many days we have left. A treewide PR from `mkPackageOptionMD` to `mkPackageOption` will probably be done at some point in 23.11, as `mkPackageOptionMD` stuff existed before this PR.

h7x4 commented:
This is ready for reviews. I don't think I'll be changing it much more apart from review comments.

If it doesn't get merged within 23.11 feature freeze, I'll swap all the `mkPackageOptionMD`s to their `mkPackageOption` variant, but for now I think I'll just keep it as is, since we'll need another treewide PR to convert the others back anyway.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2815

h7x4 committed:
treewide: use `mkPackageOption`

This commit replaces a lot of usages of `mkOption` with the package
type, to be `mkPackageOption`, in order to reduce the amount of code.

mohe2015 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -102,12 +102,7 @@ in
-        defaultText = lib.literalExpression "pkgs.mattermost";
-        description = lib.mdDoc "Mattermost derivation to use.";
-      };
+      package = mkPackageOption pkgs "mattermostl" { };
```
Typo in `mattermost`

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -190,13 +190,8 @@ in {
-        PHP package to use for Nextcloud.
-      '';
+    phpPackage = mkPackageOption pkgs "php" {
+      example = "php82";
```
`relatedPackages` and the metadata associated with it is effectively thrown out of the manual now that such an option is recommended, correct?

h7x4 replied:
Right, I forgot about these. While going through the options, I was planning of taking them into consideration later.

Using `mkPackageOption` and specifying `relatedPackages` aren't mutually exclusive - it would probably be trivial to add an extra argument to `mkPackageOption`. 

However, I'm questioning their usecase. Consider these scenarios:

- There's two competing implementations of a package (e.g. heimdal and mit kerberos) -> One is default and the other is marked as an example. No need for `relatedPackages`
- There's different versions of a package (like php above) -> We specify the non-versioned alias as default (or at least `defaultText`) and provide a versioned one as example. The user will most likely understand that there are more versions to be found. `relatedPackages` is not needed.
- There's some sort of `extensions` or `plugins` option, that takes `listOf package`, and we'd like to show possible options -> Depending on the amount of related packages we have, maintaining such a list can become hard and kinda useless at a certain point. It's easy to forget to add an entry to a nixos module everytime you add a related package. It would be better to create package sets like `tmuxPlugins`. Smaller sets might outgrow the reasonable size for `relatedPackages` at some point.

Unless I've missed something, that only leaves the case of when there's 3 or more competing implementations for a module, which seems like a rare occurence.

As a side note, these don't appear on https://search.nixos.org.

If we are to keep them, I think it would be a good idea to go through all the options and make sure to specify it properly where it's relevant, and maybe add a small note in the contribution docs.

Ma27 replied:
>  it would probably be trivial to add an extra argument to mkPackageOption.

That's what bothers me about those shortcut functions a bit: if we add a little more, we'll just have another `mkOption`.

> We specify the non-versioned alias as default (or at least defaultText) and provide a versioned one as example. The user will most likely understand that there are more versions to be found. relatedPackages is not needed.

What if you have an interpreter (such as PHP in this example), but want to denote with this option which packages are compatible (e.g. the latest one or oldest one isn't).

> There's two competing implementations of a package (e.g. heimdal and mit kerberos) -> One is default and the other is marked as an example. No need for relatedPackages

For n=2 pkgs, sure.

> Depending on the amount of related packages we have, maintaining such a list can become hard and kinda useless at a certain point. It's easy to forget to add an entry to a nixos module everytime you add a related package. It would be better to create package sets like tmuxPlugins. Smaller sets might outgrow the reasonable size for relatedPackages at some point.

Well, it's not an enum where evaluation will fail with stuff not listed in there. Also, I think that's a case where it's better to describe that drvs from a given prefix are expected rather than listing every single one of them.

> As a side note, these don't appear on https://search.nixos.org.

I assume you mean the option search not the package search that appears by default?
But anyways, it does in the man-page and the HTML, if the nixos.org stuff doesn't, then it's a bug on their end :shrug:

h7x4 replied:
> That's what bothers me about those shortcut functions a bit: if we add a little more, we'll just have another mkOption.

That's true. But most of the common boilerplate will still be gone. Maybe it would be possible to reduce the maintaince cost of `mkPackageOption` (and similar) by passing a set of defaults to `mkOption` while also allowing freeform arguments, similar to how it's usually done with special builders that use `mkDerivation`?

> What if you have an interpreter (such as PHP in this example), but want to denote with this option which packages are compatible (e.g. the latest one or oldest one isn't).

I still think this option wouldn't be a good fit. By it's name `relatedPackages`, I wouldn't jump to the conclusion that there is a version restriction. It sounds like it's better suited for a `::: {.warning}` comment inside `extraDescription`.

> I assume you mean the option search not the package search that appears by default?
> But anyways, it does in the man-page and the HTML, if the nixos.org stuff doesn't, then it's a bug on their end 🤷

Yes, I meant the option search. I feel like the fact that it's missing and have not been reported as missing yet could be an indication that this option is not that useful as things are right now. I don't know whether that's due to underutilization or because it's usecase is genuinely limited. But as stated earlier, I suspect it's the latter.

Ma27 replied:
> But most of the common boilerplate will still be gone

I disagree here, I don't consider duplication/boilerplate necessarily bad.

`package` options don't have much that I'd expect to change soonish, but on the other hand you can configure your editor to implement a small snippet.

Here you need to remember the order of the arguments (and snippets won't help as much since you'd need to write special snippets for e.g. `extraDescription` etc.) where you have a single interface for `mkOption` itself. Also, I'm wondering how prone to misuse this is, a lot of people write e.g. `mkEnableOption "Whether to allow foo"` which results in `Whether to enable Whether to allow foo.` in the description.

But anyways, the decision was made already (even before this PR) and that's not a hill I'd like to die on.

> I wouldn't jump to the conclusion that there is a version restriction. It sounds like it's better suited for a ::: {.warning} comment inside extraDescription.

Fair point.

> that this option is not that useful as things are right now.

Correct. Upon reflecting here I also came to the conclusion that I'm probably a vast minority here, so I'll resolve this now and leave things as they are.

I'll probably resume writing `package = mkOption`, others will fix it up later and that's fine for me.

Anyways, thanks a lot for replying :)
